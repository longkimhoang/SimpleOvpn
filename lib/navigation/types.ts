import {StackNavigationProp, StackScreenProps} from '@react-navigation/stack';

export type RootStackParamList = {
  'vpn-server.list': undefined;
};

export type VpnServerListScreenProps = StackScreenProps<
  RootStackParamList,
  'vpn-server.list'
>;

export type VpnServerListNavigationProps = StackNavigationProp<
  RootStackParamList,
  'vpn-server.list'
>;
