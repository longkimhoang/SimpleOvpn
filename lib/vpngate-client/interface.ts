import {IVpnServer} from './models';

export interface FetchServersOptions {
  signal?: AbortSignal;
}

export interface VpnServerRepository {
  vpnServers: readonly IVpnServer[];
  isFetching: boolean;
  fetchServers(options?: FetchServersOptions): void;
}
