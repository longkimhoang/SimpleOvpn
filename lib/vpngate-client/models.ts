import Realm, {ObjectSchema} from 'realm';

export interface IVpnServer {
  hostName: string;
  ipAddress: string;
  score: number;
  ping: number;
  speed: number;
  country: string;
  countryCode: string;
  operator: string;
  base64EncodedOvpnConfig: string;
}

export class VpnServer extends Realm.Object implements IVpnServer {
  _id!: Realm.BSON.ObjectId;
  hostName!: string;
  ipAddress!: string;
  score!: number;
  ping!: number;
  speed!: number;
  country!: string;
  countryCode!: string;
  operator!: string;
  base64EncodedOvpnConfig!: string;

  static generate(
    properties: IVpnServer,
  ): RealmInsertionModel<VpnServer> {
    return {
      ...properties,
      _id: new Realm.BSON.ObjectId(),
    };
  }

  static schema: ObjectSchema = {
    name: 'VpnServer',
    primaryKey: '_id',
    properties: {
      _id: 'objectId',
      hostName: 'string',
      ipAddress: 'string',
      score: {
        type: 'int',
        indexed: true,
      },
      ping: 'int',
      country: 'string',
      countryCode: 'string',
      operator: 'string',
      base64EncodedOvpnConfig: 'string',
    },
  };
}
