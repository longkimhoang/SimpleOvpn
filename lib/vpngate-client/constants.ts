export const VPN_GATE_SERVERS_LIST_URL = 'https://www.vpngate.net/api/iphone/';
