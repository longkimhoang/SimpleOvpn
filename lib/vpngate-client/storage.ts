import {useMMKV, useMMKVObject} from 'react-native-mmkv';
import {IVpnServer} from './models';

/**
 * A type that manages local storage of fetched VPN servers
 */
export interface VpnServerStorage {
  data: readonly IVpnServer[];
  write(data: readonly IVpnServer[]): void | Promise<void>;
}

export function useMmkvVpnServerStorage(): VpnServerStorage {
  const storage = useMMKV({
    id: 'vpngate',
  });

  const [servers, setServers] = useMMKVObject<readonly IVpnServer[]>(
    'vpn-servers',
    storage,
  );

  return {
    data: servers ?? [],
    write: setServers,
  };
}
