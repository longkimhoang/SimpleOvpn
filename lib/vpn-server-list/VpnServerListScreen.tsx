import {useNavigation} from '@react-navigation/native';
import React, {useEffect, useLayoutEffect} from 'react';
import {StyleSheet, View} from 'react-native';
import {ActivityIndicator} from 'react-native-paper';
import Animated, {useAnimatedStyle} from 'react-native-reanimated';
import NavigationBar from '../common/NavigationBar';
import {VpnServerListNavigationProps} from '../navigation/types';
import {useVpnGateClient} from '../vpngate-client';
import RefetchServersButton from './components/RefetchServersButton';
import VpnServerList from './components/VpnServerList';

const LoadingIndicatorWrapper: React.FC<{isFetching: boolean}> = ({
  isFetching,
}) => {
  const animatedStyle = useAnimatedStyle(() => {
    return {
      opacity: isFetching ? 1 : 0,
    };
  }, [isFetching]);

  return (
    <Animated.View
      pointerEvents="none"
      style={[styles.loadingIndicator, animatedStyle]}>
      <ActivityIndicator animating={isFetching} hidesWhenStopped={false} />
    </Animated.View>
  );
};

function VpnServerListScreen() {
  const {vpnServers, isFetching, fetchServers} = useVpnGateClient();
  const navigation = useNavigation<VpnServerListNavigationProps>();

  useEffect(() => {
    if (vpnServers.length > 0) {
      return;
    }

    const ac = new AbortController();

    fetchServers({signal: ac.signal});

    return () => {
      ac.abort();
    };
  }, []);

  useLayoutEffect(() => {
    navigation.setOptions({
      header: props => (
        <NavigationBar
          {...props}
          headerRight={() => (
            <RefetchServersButton fetchServers={fetchServers} />
          )}
        />
      ),
    });
  }, [navigation]);

  return (
    <View style={styles.root}>
      <LoadingIndicatorWrapper isFetching={isFetching} />

      <VpnServerList vpnServers={vpnServers} />
    </View>
  );
}

const styles = StyleSheet.create({
  root: {
    position: 'relative',
    flex: 1,
  },
  loadingIndicator: {
    position: 'absolute',
    width: '100%',
    height: '100%',
    justifyContent: 'center',
  },
});

export default VpnServerListScreen;
