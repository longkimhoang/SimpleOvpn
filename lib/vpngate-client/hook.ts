import csv from 'csvtojson';
import {useCallback, useState} from 'react';
import compactMap from '../common/compactMap';
import RealmContext from '../db/realmContext';
import {VPN_GATE_SERVERS_LIST_URL} from './constants';
import {VpnServerRepository} from './interface';
import {IVpnServer, VpnServer} from './models';

//#region Typedefs

export type FetchVpnServersAction = () => Promise<readonly IVpnServer[]>;

export interface UseVpnGateClientDeps {
  fetchVpnServersAction: FetchVpnServersAction;
}

//#endregion

//#region Fetch and parse VPNGate response

const converter = csv();

function convertToVpnServer(value: any): IVpnServer {
  return {
    hostName: value['#HostName'],
    ipAddress: value['IP'],
    score: parseInt(value['Score']),
    ping: parseInt(value['Ping']),
    speed: parseInt(value['Speed']),
    country: value['CountryLong'],
    countryCode: value['CountryShort'],
    operator: value['Operator'],
    base64EncodedOvpnConfig: value['OpenVPN_ConfigData_Base64'],
  };
}

export const fetchVpnGateServers: FetchVpnServersAction = async () => {
  const response = await fetch(VPN_GATE_SERVERS_LIST_URL);
  const body = await response.text();

  const lines = body.split(/\r?\n/).map(s => s.trim());
  const data = lines.slice(1).join('\n');

  return converter
    .fromString(data)
    .then(values => compactMap(values, convertToVpnServer));
};

//#endregion

//#region Hook

export function useVpnGateClient(
  deps: UseVpnGateClientDeps = {
    fetchVpnServersAction: fetchVpnGateServers,
  },
): VpnServerRepository {
  const {fetchVpnServersAction} = deps;
  const {useRealm, useQuery} = RealmContext;

  const [isFetching, setIsFetching] = useState(() => false);
  const realm = useRealm();
  const result = useQuery(VpnServer);
  const dispatch = useCallback(() => {
    setIsFetching(true);

    fetchVpnServersAction().then(vpnServers => {
      realm.write(() => {
        for (const server of vpnServers) {
          realm.create(VpnServer, VpnServer.generate(server));
        }
      });

      setIsFetching(false);
    });
  }, []);

  return {
    vpnServers: result,
    isFetching,
    fetchServers: dispatch,
  };
}

//#endregion
