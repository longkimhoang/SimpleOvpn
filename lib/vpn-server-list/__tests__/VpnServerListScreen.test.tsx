import {BottomSheetModalProvider} from '@gorhom/bottom-sheet';
import {BottomSheetModalMethods} from '@gorhom/bottom-sheet/lib/typescript/types';
import {fireEvent, render} from '@testing-library/react-native';
import React from 'react';
import {
  advanceAnimationByTime,
  withReanimatedTimer,
} from 'react-native-reanimated/lib/reanimated2/jestUtils';
import {VpnServerRepository} from '../../vpngate-client';
import {IVpnServer} from '../../vpngate-client/models';
import VpnServerListScreen from '../VpnServerListScreen';

describe('VpnServerListScreen', () => {
  const mockEmptyStateUseVpnServersRepository = jest.fn(() => {
    const repository: VpnServerRepository = {
      isFetching: false,
      vpnServers: [],
      fetchServers: jest.fn(),
    };

    return repository;
  });

  test('render empty state', () => {
    const mockUseVpnServerListNavigation = () => ({
      setOptions: jest.fn(),
    });

    withReanimatedTimer(() => {
      const tree = render(
        <BottomSheetModalProvider>
          <VpnServerListScreen
            overrideDeps={{
              useVpnServerRepository: mockEmptyStateUseVpnServersRepository,
              useVpnServerListNavigation: mockUseVpnServerListNavigation,
            }}
          />
        </BottomSheetModalProvider>,
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  test('render loading state', () => {
    const mockUseVpnServersRepository = jest.fn(() => {
      const repository: VpnServerRepository = {
        isFetching: true,
        vpnServers: [],
        fetchServers: jest.fn(),
      };

      return repository;
    });

    const mockUseVpnServerListNavigation = () => ({
      setOptions: jest.fn(),
    });

    withReanimatedTimer(() => {
      const tree = render(
        <BottomSheetModalProvider>
          <VpnServerListScreen
            overrideDeps={{
              useVpnServerRepository: mockUseVpnServersRepository,
              useVpnServerListNavigation: mockUseVpnServerListNavigation,
            }}
          />
        </BottomSheetModalProvider>,
      ).toJSON();

      advanceAnimationByTime(300);

      expect(tree).toMatchSnapshot();
    });
  });

  test('pass botton sheet ref correctly', () => {
    const mockUseVpnServerListNavigation = () => ({
      setOptions: jest.fn(),
    });

    type BottomSheetModalRef =
      React.MutableRefObject<BottomSheetModalMethods | null>;

    let mockBottomSheetModalRef: BottomSheetModalMethods | null = null;

    const mockUseBottomSheetModalRef = () =>
      Object.defineProperty({}, 'current', {
        get: () => mockBottomSheetModalRef,
        set: value => (mockBottomSheetModalRef = value),
      }) as BottomSheetModalRef;

    withReanimatedTimer(() => {
      render(
        <BottomSheetModalProvider>
          <VpnServerListScreen
            overrideDeps={{
              useVpnServerRepository: mockEmptyStateUseVpnServersRepository,
              useVpnServerListNavigation: mockUseVpnServerListNavigation,
              useBottomSheetModalRef: mockUseBottomSheetModalRef,
            }}
          />
        </BottomSheetModalProvider>,
      );

      expect(mockBottomSheetModalRef).toHaveProperty('present');
    });
  });

  test('when onPress is fired, bottomSheetModalRef.present will be invoked', () => {
    const mockVpnServer: IVpnServer = {
      hostName: 'public-vpn-90',
      ipAddress: `219.100.37.55`,
      score: 3037585,
      ping: 16,
      speed: 71957702,
      country: 'Japan',
      countryCode: 'JP',
      operator: 'Daiyuu Nobori_ Japan. Academic Use Only.',
      base64EncodedOvpnConfig: 'mockBase64OvpnConfig',
    };

    const mockUseVpnServersRepository = jest.fn(() => {
      const repository: VpnServerRepository = {
        isFetching: false,
        vpnServers: [mockVpnServer],
        fetchServers: jest.fn(),
      };

      return repository;
    });

    const mockUseVpnServerListNavigation = () => ({
      setOptions: jest.fn(),
    });

    const mockPresentConnectionOptionsSheet = jest.fn();

    withReanimatedTimer(() => {
      const {queryByTestId} = render(
        <BottomSheetModalProvider>
          <VpnServerListScreen
            overrideDeps={{
              useVpnServerRepository: mockUseVpnServersRepository,
              useVpnServerListNavigation: mockUseVpnServerListNavigation,
              presentConnectionOptionsSheet: mockPresentConnectionOptionsSheet,
            }}
          />
        </BottomSheetModalProvider>,
      );

      const cellElement = queryByTestId('vpn-server-list.cell');
      fireEvent(cellElement!, 'onPress');

      expect(mockPresentConnectionOptionsSheet).toBeCalledTimes(1);
      expect(mockPresentConnectionOptionsSheet).toBeCalledWith(mockVpnServer);
    });
  });

  test('given non-empty vpnServers array, when mounted, should not call fetchServers', () => {
    const mockFetchServers = jest.fn();
    const mockUseVpnServersRepository = jest.fn(() => {
      const repository: VpnServerRepository = {
        isFetching: false,
        vpnServers: [
          {
            hostName: 'public-vpn-90',
            ipAddress: `219.100.37.55`,
            score: 3037585,
            ping: 16,
            speed: 71957702,
            country: 'Japan',
            countryCode: 'JP',
            operator: 'Daiyuu Nobori_ Japan. Academic Use Only.',
            base64EncodedOvpnConfig: 'mockBase64OvpnConfig',
          },
        ],
        fetchServers: mockFetchServers,
      };

      return repository;
    });

    const mockUseVpnServerListNavigation = () => ({
      setOptions: jest.fn(),
    });

    withReanimatedTimer(() => {
      render(
        <BottomSheetModalProvider>
          <VpnServerListScreen
            overrideDeps={{
              useVpnServerRepository: mockUseVpnServersRepository,
              useVpnServerListNavigation: mockUseVpnServerListNavigation,
            }}
          />
        </BottomSheetModalProvider>,
      );

      expect(mockFetchServers).toHaveBeenCalledTimes(0);
    });
  });

  test('given empty vpnServers array, when mounted, should call fetchServers', () => {
    const mockFetchServers = jest.fn();
    const mockUseVpnServersRepository = jest.fn(() => {
      const repository: VpnServerRepository = {
        isFetching: false,
        vpnServers: [],
        fetchServers: mockFetchServers,
      };

      return repository;
    });

    const mockUseVpnServerListNavigation = () => ({
      setOptions: jest.fn(),
    });

    withReanimatedTimer(() => {
      render(
        <BottomSheetModalProvider>
          <VpnServerListScreen
            overrideDeps={{
              useVpnServerRepository: mockUseVpnServersRepository,
              useVpnServerListNavigation: mockUseVpnServerListNavigation,
            }}
          />
        </BottomSheetModalProvider>,
      );

      expect(mockFetchServers).toHaveBeenCalledTimes(1);
    });
  });

  test('when mounted, should set the header for the navigation stack', () => {
    const mockSetOptions = jest.fn();

    const mockUseVpnServerListNavigation = () => ({
      setOptions: mockSetOptions,
    });

    withReanimatedTimer(() => {
      render(
        <BottomSheetModalProvider>
          <VpnServerListScreen
            overrideDeps={{
              useVpnServerRepository: mockEmptyStateUseVpnServersRepository,
              useVpnServerListNavigation: mockUseVpnServerListNavigation,
            }}
          />
        </BottomSheetModalProvider>,
      );

      expect(mockSetOptions).toBeCalledTimes(1);

      const setOptionsArgs = mockSetOptions.mock.calls[0];
      expect(setOptionsArgs.length).toBe(1);
      expect(setOptionsArgs[0]).toHaveProperty('header');

      const headerFn = setOptionsArgs[0].header;
      expect(headerFn).toBeInstanceOf(Function);

      const {queryByTestId} = render(
        headerFn({
          options: {
            headerTitle: 'Title',
          },
        }),
      );
      expect(queryByTestId('navbar.title')).toHaveTextContent('Title');
    });
  });
});
