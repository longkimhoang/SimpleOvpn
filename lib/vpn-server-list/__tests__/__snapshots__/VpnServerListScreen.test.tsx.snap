// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VpnServerListScreen render empty state 1`] = `
Array [
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      Array [
        undefined,
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "bottom": 0,
          "overflow": "hidden",
          "top": 0,
        },
      ]
    }
  />,
  <View
    style={
      Object {
        "flex": 1,
        "position": "relative",
      }
    }
  >
    <View
      animatedStyle={
        Object {
          "value": Object {
            "opacity": 0,
          },
        }
      }
      collapsable={false}
      pointerEvents="none"
      style={
        Object {
          "height": "100%",
          "justifyContent": "center",
          "opacity": 0,
          "position": "absolute",
          "width": "100%",
        }
      }
      testID="vpn-server-list.loading-indicator.container"
    >
      <View
        accessibilityRole="progressbar"
        accessibilityState={
          Object {
            "busy": false,
          }
        }
        accessible={true}
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 24,
              "opacity": 1,
              "width": 24,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 24,
                  "transform": Array [
                    Object {
                      "rotate": "45deg",
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 12,
                    "overflow": "hidden",
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "height": 24,
                      "transform": Array [
                        Object {
                          "translateY": 0,
                        },
                        Object {
                          "rotate": "-165deg",
                        },
                      ],
                      "width": 24,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      Object {
                        "height": 12,
                        "overflow": "hidden",
                        "width": 24,
                      }
                    }
                  >
                    <View
                      collapsable={false}
                      style={
                        Object {
                          "borderColor": "rgba(103, 80, 164, 1)",
                          "borderRadius": 12,
                          "borderWidth": 2.4,
                          "height": 24,
                          "width": 24,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            collapsable={false}
            style={
              Object {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 24,
                  "transform": Array [
                    Object {
                      "rotate": "45deg",
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 12,
                    "overflow": "hidden",
                    "top": 12,
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "height": 24,
                      "transform": Array [
                        Object {
                          "translateY": -12,
                        },
                        Object {
                          "rotate": "345deg",
                        },
                      ],
                      "width": 24,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      Object {
                        "height": 12,
                        "overflow": "hidden",
                        "width": 24,
                      }
                    }
                  >
                    <View
                      collapsable={false}
                      style={
                        Object {
                          "borderColor": "rgba(103, 80, 164, 1)",
                          "borderRadius": 12,
                          "borderWidth": 2.4,
                          "height": 24,
                          "width": 24,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
    <RCTScrollView
      collapsable={false}
      data={Array []}
      getItem={[Function]}
      getItemCount={[Function]}
      getItemLayout={[Function]}
      handlerTag={1}
      handlerType="NativeViewGestureHandler"
      initialNumToRender={8}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      renderScrollComponent={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Array [
          Object {
            "backgroundColor": "rgba(255, 251, 254, 1)",
          },
        ]
      }
      viewabilityConfigCallbackPairs={Array []}
      windowSize={5}
    >
      <View />
    </RCTScrollView>
  </View>,
]
`;

exports[`VpnServerListScreen render loading state 1`] = `
Array [
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      Array [
        undefined,
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "bottom": 0,
          "overflow": "hidden",
          "top": 0,
        },
      ]
    }
  />,
  <View
    style={
      Object {
        "flex": 1,
        "position": "relative",
      }
    }
  >
    <View
      animatedStyle={
        Object {
          "value": Object {
            "opacity": 1,
          },
        }
      }
      collapsable={false}
      pointerEvents="none"
      style={
        Object {
          "height": "100%",
          "justifyContent": "center",
          "opacity": 1,
          "position": "absolute",
          "width": "100%",
        }
      }
      testID="vpn-server-list.loading-indicator.container"
    >
      <View
        accessibilityRole="progressbar"
        accessibilityState={
          Object {
            "busy": true,
          }
        }
        accessible={true}
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 24,
              "opacity": 1,
              "width": 24,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 24,
                  "transform": Array [
                    Object {
                      "rotate": "45deg",
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 12,
                    "overflow": "hidden",
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "height": 24,
                      "transform": Array [
                        Object {
                          "translateY": 0,
                        },
                        Object {
                          "rotate": "-165deg",
                        },
                      ],
                      "width": 24,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      Object {
                        "height": 12,
                        "overflow": "hidden",
                        "width": 24,
                      }
                    }
                  >
                    <View
                      collapsable={false}
                      style={
                        Object {
                          "borderColor": "rgba(103, 80, 164, 1)",
                          "borderRadius": 12,
                          "borderWidth": 2.4,
                          "height": 24,
                          "width": 24,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            collapsable={false}
            style={
              Object {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 24,
                  "transform": Array [
                    Object {
                      "rotate": "45deg",
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 12,
                    "overflow": "hidden",
                    "top": 12,
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "height": 24,
                      "transform": Array [
                        Object {
                          "translateY": -12,
                        },
                        Object {
                          "rotate": "345deg",
                        },
                      ],
                      "width": 24,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      Object {
                        "height": 12,
                        "overflow": "hidden",
                        "width": 24,
                      }
                    }
                  >
                    <View
                      collapsable={false}
                      style={
                        Object {
                          "borderColor": "rgba(103, 80, 164, 1)",
                          "borderRadius": 12,
                          "borderWidth": 2.4,
                          "height": 24,
                          "width": 24,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
    <RCTScrollView
      collapsable={false}
      data={Array []}
      getItem={[Function]}
      getItemCount={[Function]}
      getItemLayout={[Function]}
      handlerTag={2}
      handlerType="NativeViewGestureHandler"
      initialNumToRender={8}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      renderScrollComponent={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Array [
          Object {
            "backgroundColor": "rgba(255, 251, 254, 1)",
          },
        ]
      }
      viewabilityConfigCallbackPairs={Array []}
      windowSize={5}
    >
      <View />
    </RCTScrollView>
  </View>,
]
`;
